File: problem1.cpp
// Time Complexity : O(n) --> We iterate throught the entire array once
// Space Complexity : O(1) --> Only using the output array
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Reducing space complexity from O(n) to O(1)


// Your code here along with comments explaining your approach

/*
1. Create a temporary pointer that checks for all the values to left of the original array.  
2. After appending the left hand side product, iterate from end to beginning updating the values for right hand side products
*/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& arr) {
      
        int vectorSize = arr.size();
        vector<int> answer(vectorSize);
        
        int temp = 1;
        
  
        for(int i = 0; i < vectorSize; i++)
        {
            answer[i] = temp;
            temp *= arr[i];
        }
        
		temp = 1;
     
        for(int i = vectorSize - 1;  i >= 0; i-- )
        {
            answer[i] = answer[i] * temp;
            temp *= arr[i];
        }
        

		return answer;

	}
    
};

File: problem_1.java
// Time Complexity :O(n)
// Space Complexity :Auxiliary O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
import java.util.Arrays;
class problem_1 {
    public int[] productExceptSelf(int[] nums) {
        if(nums == null || nums.length == 0) return nums;
        int n = nums.length;
        int [] result = new int[n];
        int rp = 1;
        result[0] = 1;
        // l to r
        for(int i =1; i < n; i++){
            rp *= nums[i-1];
            result[i]= rp;
        }
        // r to l
        rp = 1;
        for(int i = n-2; i >= 0; i--){
            rp *= nums[i+1];
            result[i] *= rp;
        }
        return result;
    }

    public static void main(String args[]){
        problem_1 obj = new problem_1();
        // test case 1
        int [] nums = {1,2,3,4};
        int [] output = obj.productExceptSelf(nums);
        System.out.println(Arrays.toString(output));
    }
}